### DISCUSSION TITLE
Why can't i set both region and endpointConfiguration?

### DISCUSSION THREAD

**Discussion by alin-simionoiuDE** [aws/aws-sdk-java#2818] [Q&A]
The code that enforces this is in `AwsClientBuilder.setRegion`. I don't understand this limitation, why can't i set both endpoint configuration and region?

This question comes as a result of a weird situation when i use [localstack](https://github.com/localstack/localstack)

When i set the endpoint to point to localstack (which runs kinesis) AND the client region is set to `us-west-2`, the client shows this error

````
 com.amazonaws.services.kinesis.model.ResourceNotFoundException: Stream arn arn:aws:kinesis:us-west-2:000000000000:stream/testing-stream not found (Service: AmazonKinesis; Status Code: 400; Error Code: ResourceNotFoundException; Request ID: 7335965c-142e-11ed-b38c-55a989fea18a; Proxy: null)
````

while `localstack` complains that

```
kinesis.mock.ResourceNotFoundException: Stream arn arn:aws:kinesis:us-east-1:000000000000:stream/testing-stream not found
```

Which means the client is looking for a stream where the ARN has `us-west-2` but what is asking for is a stream where the ARN has `us-east-1`. i debugged the java code and is tracking.

this is really puzzling. again, why can't i set both region and endpoint? or there's a way to do this ? maybe without using the builder

**Comment by yasminetalby** [comment]
Hello @alins ,

Thank you very much for your message.
Could you please provide a code sample to review?
Have you attempted the following syntax (with appropriate serviceEndpoint, serviceRegion)?:

`.withEndpointConfiguration(new EndpointConfiguration(
          String serviceEndpoint,
        String serviceRegion))`

Best,
Yasmine

**Comment by alin-simionoiuDE** [comment]
Hello @yasminetalby,

Thank you so much for the reply. Yes, I am using `withEndpointConfiguration` when I create an `AmazonKinesis`, my problem is when I configure a worker

so when I do something like

```
        Worker.Builder builder = new Worker.Builder()
            .recordProcessorFactory(...)
            .config(config);
```

`config` is of type `KinesisClientLibConfiguration`, which doesn't have a setter for `EndpointConfiguration` type, only a simple 

````
    withKinesisEndpoint(String endpoint)
````

all this means, that I will endpoint into a situation where I set the endpoint for the worker (thru `KinesisClientLibConfiguration`) and the signing region is gonna fall to the default `us-east-1`.

i hope this makes sense

**Reply by yasminetalby** [reply]
Hello @alins ,

I sincerely apologies for the delay.
It seems that you are in the wrong repository. This is the repository that supports https://github.com/awslabs/amazon-kinesis-client
They have some public documentation that might be helpful for your use case in the developer guide: https://docs.aws.amazon.com/streams/latest/dev/kinesis-record-processor-implementation-app-java.html#kcl-java-worker

Best,

Yasmine

**Comment by alin-simionoiuDE** [comment]
btw, in the newer versions of aws SDK `EndpointConfiguration` is completely gone. i see only something like

```
B endpointOverride(URI endpointOverride);
```

how do I set signing region in this case?

**Reply by yasminetalby** [reply]
Hello @alins ,
Regarding the change in endpointOverride between AWS Java SDK v1 and v2, here is how to provide the signing region:
```
.builder()
.region(signingRegion)
.endpointOverride(endpoint)
.build()
```
You can look at the [ changelog documentation](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md#12-aws-client-configuration-custom-regions-credentials-and-endpoints) for more informations regarding changes between AWS Java SDK v1 and v2.

Best regards,

Yasmine
